fedimint-client/src/meta.rs

```diff
@@ -0,0 +1,262 @@
+use std::collections::BTreeMap;
+use std::sync::Arc;
+use std::time::Duration;
+
+use anyhow::{bail, Context as _};
+use fedimint_core::db::{Database, IDatabaseTransactionOpsCoreTyped};
+use fedimint_core::task::waiter::Waiter;
+use fedimint_core::task::{MaybeSend, MaybeSync};
+use fedimint_core::util::{retry, FibonacciBackoff};
+use fedimint_core::{apply, async_trait_maybe_send};
+use serde::de::DeserializeOwned;
+use tracing::{debug, instrument, warn};
+
+use crate::db::{MetaFieldKey, MetaFieldPrefix, MetaFieldValue};
+use crate::Client;
+
+#[apply(async_trait_maybe_send!)]
+pub trait MetaSource: MaybeSend + MaybeSync + 'static {
+    /// Wait for next change in this source.
+    async fn wait_for_update(&self);
+    async fn fetch(
+        &self,
+        client: &Client,
+        fetch_kind: FetchKind,
+    ) -> anyhow::Result<BTreeMap<MetaFieldKey, MetaFieldValue>>;
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum FetchKind {
```

Comment:
Why make it an implementation concern? Can't caller of `fetch` deal with retrying?

Reply:
if you have two network calls in fetch, one succeeded you don't want to retry that one. it is consequence of having just a single meta source